suppress Traceback & AssertionError for plugin errors

Currently, the CLI prints the full Traceback and AssertionError when a package is incompatible with napari.

## What I get

When purposefully mis-spelling `Framework :: napari` to be `Framework :: napar`, I get the following when trying to validate the package...

```bash
(base) √ napari-hdf5-labels-io % npd --validate-package  

Traceback (most recent call last):
  File "/Users/justin.kiggins/opt/anaconda3/bin/npd", line 8, in <module>
    sys.exit(main())
  File "/Users/justin.kiggins/opt/anaconda3/lib/python3.8/site-packages/napari_plugin_devtools/__main__.py", line 25, in main
    validate_packages('dist')
  File "/Users/justin.kiggins/opt/anaconda3/lib/python3.8/site-packages/napari_plugin_devtools/validation.py", line 36, in validate_packages
    assert (
AssertionError: Classifier Framework :: napari must be specified for the plugin to be discovered
```

Displaying the Traceback here is confusing because it seems like there is an error in the CLI tool itself


## What I expected

```bash
(base) √ napari-hdf5-labels-io % npd --validate-package  
Woops! You must specify "Framework :: napari" as a classifier for napari to recognize this package as containing napari plugins, but it was not found. 

The following classifiers were found:

        'Development Status :: 4 - Beta',
        'Intended Audience :: Developers',
        'Framework :: napar',
        'Topic :: Software Development :: Testing',
        'Programming Language :: Python',
        'Programming Language :: Python :: 3',
        'Programming Language :: Python :: 3.7',
        'Programming Language :: Python :: 3.8',
        'Programming Language :: Python :: 3.9',
        'Operating System :: OS Independent',
        'License :: OSI Approved :: GNU General Public License v3 (GPLv3)',

Update your classifiers to include "Framework :: napari", then rebuild the package.

```


## A proposed solution
I would recommend the following...

1. capture AssertionErrors with try/except blocks
2. where possible provide additional guidance and context to the user, as in the example above
3. use `exit(1)` or a more appropriate exit code (is there a better one?) to exit so that CI/CD pipelines still throw an error

--- a/README.md
+++ b/README.md
@@ -7,6 +7,11 @@
 
 A repo with tools and services for [napari plugin](https://napari.org/docs/dev/plugins/index.html) developers
 
+## Installation
+This tool can be installed via pip: `pip install napari-plugin-devtools`, 
+you can also add it to your requirements.txt to be installed with other 
+dependencies automatically.
+
 ## Validation tool
 The validation tools provides automated tests for plugin developers, 
 these checks verify that the plugin is available for users of napari 
@@ -29,11 +34,11 @@ checks that all packages built under dist are correctly tagged with
 classifier "Framework :: napari". This is recommended for most plugins
 unless you do not want your plugin to appear in the napari built-in plugin installation tool.
 
-`npd --validate-functions` verifies that after the plugin is installed 
-to current python environment, typically done through `pip install -e .`, 
-there is at least one analysis function registered by the plugin, and 
-there are no conflicts in name or registration error, it also output 
-all registered function signaturesin a json format that can be used 
+`npd --validate-readers|--validate-writers|--validate-functions|--validate-widgets`, 
+verifies that after the plugin is installed to current python environment, typically done 
+through `pip install -e .`, there is at least one hook under corresponding category 
+registered by the plugin, and there are no conflicts in name or registration error, 
+it also output all registered function signatures in a json format that can be used 
 for further inspection, for example:
 ```
     [{
@@ -51,7 +56,7 @@ for further inspection, for example:
 ``` 
 
 Multiple cmds in one execution is supported, for example: 
-`npd --validate-functions --validate-packages`
+`npd --validate-readers --validate-functions --validate-packages`
 
 ### Pytest fixture usage
 devtools provides a pytest fixture: napari_plugin_tester 

--- a/napari_plugin_devtools/__init__.py
+++ b/napari_plugin_devtools/__init__.py
@@ -1,11 +1,11 @@
 __all__ = [
     "validate_packages",
     "validate_function",
-    "list_function_implementations",
+    "list_hook_implementations",
 ]
 
 from .validation import (
-    list_function_implementations,
+    list_hook_implementations,
     validate_function,
     validate_packages,
 )

--- a/napari_plugin_devtools/__main__.py
+++ b/napari_plugin_devtools/__main__.py
@@ -4,10 +4,14 @@ napari plugin command line developer tool.
 import argparse
 import sys
 
-from .validation import list_function_implementations, validate_packages
+from .validation import Options, list_hook_implementations, validate_packages
+
+print_help = True
 
 
 def main():
+    global print_help
+
     parser = argparse.ArgumentParser()
     parser.add_argument(
         '--validate-packages',
@@ -15,17 +19,56 @@ def main():
         help='validate the build packages is marked as napari plugin',
     )
     parser.add_argument(
+        '--validate-readers',
+        action="store_true",
+        help='validate there is a reader hook implemented',
+    )
+    parser.add_argument(
+        '--validate-writers',
+        action="store_true",
+        help='validate there is a writer hook implemented',
+    )
+    parser.add_argument(
         '--validate-functions',
         action="store_true",
-        help='validate the build packages is marked as napari plugin',
+        help='validate there is a function hook implemented',
+    )
+    parser.add_argument(
+        '--validate-widgets',
+        action="store_true",
+        help='validate there is a widget hook implemented',
     )
     args = parser.parse_args()
 
+    err_code = 0
     if args.validate_packages:
-        validate_packages('dist')
+        err_code += validate_packages('dist')
+        print_help = False
+
+    if args.validate_readers:
+        err_code += validate_hook(Options.reader)
+
+    if args.validate_writers:
+        err_code += validate_hook(Options.writer)
 
     if args.validate_functions:
-        print(list_function_implementations())
+        err_code += validate_hook(Options.function)
+
+    if args.validate_widgets:
+        err_code += validate_hook(Options.widget)
+
+    if print_help:
+        parser.print_help()
+    return err_code
+
+
+def validate_hook(option):
+    global print_help
+
+    err_code, signature = list_hook_implementations(option)
+    print(f'{option.name}: {signature}')
+    print_help = False
+    return err_code
 
 
 if __name__ == '__main__':

--- a/napari_plugin_devtools/validation.py
+++ b/napari_plugin_devtools/validation.py
@@ -1,14 +1,30 @@
+import enum
 import inspect
 import json
 import os
 from types import FunctionType
-from typing import Callable, Dict, List, Tuple, Union
+from typing import Callable, Dict, List, Union
 
 from napari.plugins import plugin_manager
 from napari_plugin_engine import HookImplementation
 from pkginfo import SDist, Wheel
 
-functions: Dict[Tuple[str, str], Callable] = dict()
+
+class Options(enum.Enum):
+    """A set of valid arithmetic operations for image_arithmetic."""
+
+    reader = [plugin_manager.hook.napari_get_reader]
+    writer = [
+        plugin_manager.hook.napari_get_writer,
+        plugin_manager.hook.napari_write_image,
+        plugin_manager.hook.napari_write_labels,
+        plugin_manager.hook.napari_write_points,
+        plugin_manager.hook.napari_write_shapes,
+        plugin_manager.hook.napari_write_surface,
+        plugin_manager.hook.napari_write_vectors,
+    ]
+    function = [plugin_manager.hook.napari_experimental_provide_function]
+    widget = [plugin_manager.hook.napari_experimental_provide_dock_widget]
 
 
 def validate_packages(folder):
@@ -19,11 +35,12 @@ def validate_packages(folder):
     folder : str
         dist folder after build with setuptools
 
-    Raises
+    Returns
     ------
-    AssertionError
-        If the package is not annotated properly
+    err_code : int
+        0 if every package is valid, 1 otherwise
     """
+    err_code = 0
     for pkgpath in os.listdir(folder):
         pkgpath = os.path.join(folder, pkgpath)
         if pkgpath.endswith('.tar.gz'):
@@ -33,20 +50,28 @@ def validate_packages(folder):
         else:
             print(f'Not a valid format {pkgpath}')
             continue
-        assert (
+        if not (
             hasattr(dist, 'classifiers')
             and 'Framework :: napari' in dist.classifiers
-        ), (
-            'Classifier Framework :: napari must be specified '
-            'for the plugin to be discovered'
-        )
-
-        print(f'validated {pkgpath}')
+        ):
+            err_code = 1
+            classifiers = '\n\t'.join(dist.classifiers)
+            print(
+                f'Woops! "Framework :: napari" was not found in {pkgpath}\n'
+                f'Know classifiers:\n'
+                f'\t{classifiers}\n'
+                f'Please update your package setup configuration to include '
+                f'"Framework :: napari", then rebuild the distribution.\n'
+            )
+        else:
+            print(f'validated {pkgpath}')
+    return err_code
 
 
 def validate_function(
     args: Union[Callable, List[Callable]],
     hookimpl: HookImplementation,
+    functions: Dict,
 ):
     """Validate the function is properly implemented in napari supported format
 
@@ -56,48 +81,64 @@ def validate_function(
         function to validate
     hookimpl : HookImplementation
         implementation of the hook
+    functions: Dict
+        registered functions
 
-    Raises
+    Returns
     ------
-    AssertionError
-        When function is not in supported format
+    err_code: int
+        0 if the function is valid, 1 otherwise
     """
+    err_code = 0
     plugin_name = hookimpl.plugin_name
     hook_name = '`napari_experimental_provide_function`'
+    if plugin_name is None:
+        err_code = 1
+        print("No plugin name specified")
 
-    assert plugin_name is not None, "No plugin name specified"
     for func in args if isinstance(args, list) else [args]:
-        assert not isinstance(
-            func, tuple
-        ), "To provide multiple functions please use a LIST of callables"
-        assert isinstance(func, FunctionType), (
-            f'Plugin {plugin_name!r} provided a non-callable type to '
-            f'{hook_name}: {type(func)!r}. Function ignored.'
-        )
+        if isinstance(func, tuple):
+            err_code = 1
+            print(
+                "To provide multiple functions please use a LIST of callables"
+            )
+        elif not isinstance(func, FunctionType):
+            err_code = 1
+            print(
+                f'Plugin {plugin_name!r} provided a non-callable type to '
+                f'{hook_name}: {type(func)!r}. Function ignored.'
+            )
 
         # Get function name
         name = func.__name__.replace('_', ' ')
 
         key = (plugin_name, name)
-        assert key not in functions, (
-            "Plugin '{}' has already registered a function '{}' "
-            "which has now been overwritten".format(*key)
-        )
-
+        if key in functions:
+            err_code = 1
+            print(
+                f"Plugin '{plugin_name}' has already registered a function "
+                f"'{name}' which has now been overwritten"
+            )
         functions[key] = func
 
+    return err_code
 
-def list_function_implementations(plugin_name=None):
+
+def list_hook_implementations(option, plugin_name=None):
     """
-    List function implementations found when loaded by napari
+    List hook implementations found when loaded by napari
 
     Parameters
     ----------
+    option : type of the function hook to check
     plugin_name : if provided, only show functions from the given plugin name
 
     Returns
     -------
-    json string for the list of function signatures
+    err_code: int
+        0 if there is one hook registered and no conflicts found, 1 otherwise
+    functions_signatures: str
+        Json string for the list of function signatures
 
     Example:
     [{
@@ -113,24 +154,48 @@ def list_function_implementations(plugin_name=None):
         "defaults": null
     }]
     """
-    fw_hook = plugin_manager.hook.napari_experimental_provide_function
-    functions.clear()
-    fw_hook.call_historic(result_callback=validate_function, with_impl=True)
     function_signatures = []
-    for key, value in functions.items():
-        spec = inspect.getfullargspec(value)
-        if plugin_name is None or plugin_name == key[0]:
-            function_signatures.append(
-                {
-                    'plugin name': key[0],
-                    'function name': key[1],
-                    'args': spec.args,
-                    'annotations': spec.annotations,
-                    'defaults': spec.defaults,
-                }
-            )
-    assert len(function_signatures) > 0, (
-        'No function found under the entrypoint, annotate plugin function '
-        'and add the module to the entry_points under napari.plugin'
-    )
-    return json.dumps(function_signatures, default=str)
+    err_code = 0
+    functions = dict()
+
+    if option == Options.function or option == Options.widget:
+        fw_hook = option.value[0]
+        res = fw_hook._hookexec(fw_hook, fw_hook.get_hookimpls(), None)
+        for result, impl in zip(res.result, res.implementation):
+            err_code += validate_function(result, impl, functions)
+        for key, value in functions.items():
+            spec = inspect.getfullargspec(value)
+            if plugin_name is None or plugin_name == key[0]:
+                function_signatures.append(
+                    {
+                        'plugin name': key[0],
+                        'function name': key[1],
+                        'args': spec.args,
+                        'annotations': spec.annotations,
+                        'defaults': spec.defaults,
+                    }
+                )
+    elif option == Options.reader or option == Options.writer:
+        for hook_spec in option.value:
+            for hook_impl in hook_spec.get_hookimpls():
+                if (
+                    hook_impl.plugin_name != 'builtins'
+                    and hook_impl.plugin_name != 'svg'
+                ):  # remove plugins installed by default
+                    function_signatures.append(
+                        {
+                            'plugin name': hook_impl.plugin_name,
+                            'spec': hook_impl.specname,
+                            'tryfirst': hook_impl.tryfirst,
+                            'trylast': hook_impl.trylast,
+                        }
+                    )
+
+    if len(function_signatures) == 0:
+        print(
+            f'No hook found under "{option.name}", annotate plugin hooks '
+            'and add the module to the entry_points under napari.plugin'
+        )
+        err_code = 1
+
+    return err_code, json.dumps(function_signatures, default=str)

